asmcomp/amd64/emit.mlp

```diff
@@ -542,13 +542,19 @@ let emit_instr fallthrough i =
         | _ -> I.mov (reg src) (reg dst)
         end
   | Lop(Iconst_int n) ->
+      (* use a 32bit register when we can *)
       if n = 0n then begin
         match i.res.(0).loc with
+        | Reg r when r < 13 -> I.xor (res32 i 0) (res32 i 0)
         | Reg _ -> I.xor (res i 0) (res i 0)
         | _     -> I.mov (int 0) (res i 0)
       end
-      else
-        I.mov (nat n) (res i 0)
+      else begin
+        match i.res.(0).loc with
+        | Reg r when r < 13 && n > 0x0n && n <= 0x7FFF_FFFFn
```

Comment:
I suspect you can widen the interval to `n <= 0xFFFF_FFFFn`.

Reply:
Also, why `r < 13` ?  Integer registers available for register allocation are numbered 0 to 12, so a register number above 13 would be a FP register...

Reply:
Agree with both points and have fixed.